src/libcore/iter/iterator.rs

```diff
@@ -2251,6 +2251,32 @@ pub trait Iterator {
         Product::product(self)
     }
 
+    /// Eagerly consume the iterator, evaluating all the elements.
+    ///
+    /// This is primarily useful for evaluating the effects of an iterator that
+    /// has previously been created. To evaluate the effects of an iterator
+    /// immediately, it is likely better to use [`for_each`].
+    ///
+    /// [`for_each`]: #method.for_each
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// // Prepare an iterator with side effects...
+    /// let count_aloud = 1..=10.map(|x| println!("{}", x));
+    ///
+    /// // ...
+    ///
+    /// // Trigger the effects!
+    /// count_aloud.exhaust();
+    ///
+    /// ```
+    #[inline]
+    #[unstable(feature = "iter_exhaust", issue = "44546")]
+    fn exhaust(self) where Self: Sized {
+        for _ in self {}
```

Comment:
This should use `for_each` to get internal iteration, rather than (implicit) `next()` calls.

Reply:
The difference is that some iterators can apply a `fold` much more efficiently than they can `next` one item at a time.  For instance, `Chain::fold` can fold its first part as a whole, then its second part as a whole, but `Chain::next` has to check whether it's on the first or second part every time.  This is demonstrated in the benchmarks added to the `for_each` PR, #42782, in commit 4a8ddac99e1edfb219e11c3ea2d6c43ccecb29ab.